go-controller/pkg/node/controllers/upgrade/node_upgrade.go

```diff
@@ -0,0 +1,205 @@
+
+	// Holds the initially found version
+	initialTopoVersion int
+}
+
+// NewController creates a new upgrade controller
+func NewController(client kube.Interface, nodeInformer coreinformers.NodeInformer) *upgradeController {
+	uc := &upgradeController{
+		client: client,
+		queue:  workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName),
+	}
+
+	klog.Info("Setting up event handlers for node upgrade")
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    uc.onNodeAdd,
+		UpdateFunc: uc.onNodeUpdate,
+	})
+	uc.nodeLister = nodeInformer.Lister()
+	uc.nodesSynced = nodeInformer.Informer().HasSynced
+	var err error
+	uc.initialTopoVersion, err = uc.detectInitialVersion()
+	if err != nil {
+		klog.Fatalf("Unable to run initial version detection: %v", err)
+	}
+	return uc
+}
+
+// onNodeAdd queues the Node for processing.
+func (uc *upgradeController) onNodeAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+	klog.V(4).Infof("Adding node %s", key)
+	uc.queue.Add(key)
+}
+
+// onNodeUpdate updates the Node Selector in the cache and queues the Node for processing.
+func (uc *upgradeController) onNodeUpdate(oldObj, newObj interface{}) {
+	oldNode := oldObj.(*v1.Node)
+	newNode := newObj.(*v1.Node)
+
+	// don't process resync or objects that are marked for deletion
+	if oldNode.ResourceVersion == newNode.ResourceVersion ||
+		!newNode.GetDeletionTimestamp().IsZero() {
+		return
+	}
+
+	key, err := cache.MetaNamespaceKeyFunc(newObj)
+	if err == nil {
+		uc.queue.Add(key)
+	}
+}
+
+func (uc *upgradeController) Run(stopCh <-chan struct{}, informerStop chan struct{}) error {
+	defer utilruntime.HandleCrash()
+	defer uc.queue.ShutDown()
+
+	klog.Infof("Starting controller %s", controllerName)
+	defer klog.Infof("Shutting down controller %s", controllerName)
+
+	// Wait for the caches to be synced
+	klog.Info("Waiting for informer caches to sync")
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, uc.nodesSynced) {
+		return fmt.Errorf("error syncing cache")
+	}
+
+	ticker := time.NewTicker(1 * time.Second)
+	deadline := time.After(30 * time.Minute)
+	klog.Info("Waiting for OVN Masters to upgrade to switchover K8S Service route")
+	for {
+		select {
+		case <-ticker.C:
+			if uc.processNextWorkItem() {
+				klog.Infof("Masters have completed upgrade from topology version %d to %d",
+					uc.initialTopoVersion, ovntypes.OvnCurrentTopologyVersion)
+				ticker.Stop()
+				informerStop <- struct{}{}
+				return nil
+			}
+		case <-deadline:
+			ticker.Stop()
+			informerStop <- struct{}{}
+			klog.Fatal("Failed to detect completion of master upgrade after 30 minutes. Check if " +
+				"ovnkube-masters have upgraded correctly!")
+		case <-stopCh:
+			ticker.Stop()
+			informerStop <- struct{}{}
+			return nil
+		}
+	}
+}
```

Comment:
I think that we don't need to implement a controller for this case because:
- the node only cares about itself,  are we interested in knowing about other nodes or do we depend on the apiserver?
- the ovnkube-node runs as a pod inside a daemonset that will restart it if it fails

I think that the approach of "I run normally and keep listening in the background, if I detect the master change I upgrade" is correct, but we can go simpler here with:
1. start normally
2. run this loop in the background
3. if I detect the change just exist and let the pod restart , we are testing that we can restart without disruption and we can leverage it

Reply:
scratch the comment, I talked offline and I was getting it wrong

Reply:
my point is that we can get rid of the workqueue, informer and event handles, and use the `toolswatch.UntilWtihSync` method, that will provide us the functionality that we want "want for the master node to have the annotation X"
This is basically a subset of an informer, it creates a Watcher to the apiserver for the events we want to watch, it also handles the reconnections ... the alternatives is to poll the apiserver continuously that is less efficient.
It will be something like:
```
func detectUpgradeDone(client clientset.Interfaceg) (*v1.Node, error) {
	ctx, cancelFunc := context.WithTimeout(context.TODO(), timeoutForMasterAnnotation)
	defer cancelFunc()

       // add our selector here for master nodes this 
	fieldSelector := fields.OneTermEqualSelector("metadata.name", nodeName).String()
	lw := &cache.ListWatch{
		ListFunc: func(options metav1.ListOptions) (object runtime.Object, e error) {
			options.FieldSelector = fieldSelector
			return client.CoreV1().Nodes().List(ctx, options)
		},
		WatchFunc: func(options metav1.ListOptions) (i watch.Interface, e error) {
			options.FieldSelector = fieldSelector
			return client.CoreV1().Nodes().Watch(ctx, options)
		},
	}
	condition := func(event watch.Event) (bool, error) {
		if n, ok := event.Object.(*v1.Node); ok {
				topoVer, ok := node.Annotations[ovntypes.OvnK8sTopoAnno]
	if ok && len(topoVer) > 0 {
		ver, err := strconv.Atoi(topoVer)
		if err != nil {
			klog.Warningf("Illegal value detected for %s, on node: %s, value: %s, error: %v",
				ovntypes.OvnK8sTopoAnno, node.Name, topoVer, err)
			return false, err
		}
		if ver == ovntypes.OvnCurrentTopologyVersion {
			return true, nil
		}
	}
		}
		return false, fmt.Errorf("event object not of type Node")
	}

	evt, err := toolswatch.UntilWithSync(ctx, lw, &v1.Node{}, nil, condition)
	if err != nil {
		return nil, fmt.Errorf("timeout waiting for master upgrade annotation")
	}
	if n, ok := evt.Object.(*v1.Node); ok {
		return n, nil
	}
	return nil, fmt.Errorf("event object not of type node")
}

```

this is exactly how kube-proxy wants for nodes to get the pod-cidr annotation https://github.com/kubernetes/kubernetes/blob/5bed8d355f824fcce6ef45c9dc77020189329de9/cmd/kube-proxy/app/server_others.go#L391-L423

Reply:
After discussing offline this seems better from a reduction of code perspective. I'll give it a try and if it works use it.

Reply:
I decided to hold off on this, the upgrade currently works and I dont have time to fix this. This is mostly a code optimization which we can do in a follow up.